src/V3SchedPartition.cpp

```diff
@@ -0,0 +1,386 @@
+                    new V3GraphEdge{m_graphp, varVtxp, vtxp, 1};
+                });
+
+                // Store back to hash map so we can find it next time
+                it = m_senVertices.emplace(*senItemp, vtxp).first;
+            }
+
+            // Cache sensitivity vertex
+            senItemp->user1p(it->second);
+        }
+        return senItemp->user1u().to<SchedSenVertex*>();
+    }
+
+    void visitLogic(AstNode* nodep) {
+        UASSERT_OBJ(m_senTreep, nodep, "Should be under AstActive");
+
+        SchedLogicVertex* const logicVtxp
+            = new SchedLogicVertex{m_graphp, m_scopep, m_senTreep, nodep};
+
+        // Clocked or hybrid logic has explicit sensitivity, so add edge from sensitivity vertex
+        if (!m_senTreep->hasCombo()) {
+            m_senTreep->foreach<AstSenItem>([=](AstSenItem* senItemp) {
+                UASSERT_OBJ(senItemp->isClocked() || senItemp->isHybrid(), nodep,
+                            "Non-clocked SenItem under clocked SenTree");
+                V3GraphVertex* const eventVtxp = getSenVertex(senItemp);
+                new V3GraphEdge{m_graphp, eventVtxp, logicVtxp, 10};
+            });
+        }
+
+        // Add edges based on references
+        nodep->foreach<AstVarRef>([=](const AstVarRef* vrefp) {
+            AstVarScope* const vscp = vrefp->varScopep();
+            SchedVarVertex* const varVtxp = getVarVertex(vscp);
+            if (vrefp->access().isReadOrRW() && m_readTriggersThisLogic(vscp)) {
+                new V3GraphEdge{m_graphp, varVtxp, logicVtxp, 10};
+            }
+            if (vrefp->access().isWriteOrRW()) {
+                new V3GraphEdge{m_graphp, logicVtxp, varVtxp, 10};
+            }
+        });
+
+        // If the logic calls a DPI import, it might fire the DPI Export trigger
+        if (AstVarScope* const dpiExporTrigger = v3Global.rootp()->dpiExportTriggerp()) {
+            nodep->foreach<AstCCall>([=](const AstCCall* callp) {
+                if (!callp->funcp()->dpiImportWrapper()) return;
+                SchedVarVertex* const varVtxp = getVarVertex(dpiExporTrigger);
+                new V3GraphEdge{m_graphp, logicVtxp, varVtxp, 10};
+            });
+        }
+    }
+
+    // VISIT methods
+    virtual void visit(AstActive* nodep) override {
+        AstSenTree* const senTreep = nodep->sensesp();
+        UASSERT_OBJ(senTreep->hasClocked() || senTreep->hasCombo() || senTreep->hasHybrid(), nodep,
+                    "Unhandled");
+        UASSERT_OBJ(!m_senTreep, nodep, "Should not nest");
+
+        // Mark explicit sensitivities as not triggering these blocks
+        if (senTreep->hasHybrid()) {
+            AstNode::user2ClearTree();
+            senTreep->foreach<AstVarRef>([](const AstVarRef* refp) {  //
+                refp->varScopep()->user2(true);
+            });
+        }
+
+        m_senTreep = senTreep;
+        iterateChildrenConst(nodep);
+        m_senTreep = nullptr;
```

Comment:
```suggestion
        VL_RESTORER(m_senTreep);
        m_senTreep = senTreep;
        iterateChildrenConst(nodep);
```

Not that Active will nest, but helps document that this really is being set just for recursion purposes.

Reply:
I really don't like this as it suggests it *can* nest.

Can we add a new flavour of VL_RESTORER that takes a second argument, which is the expected value, and asserts it?

```
VL_RESTORER(foo, nullptr)
```
is the same as
```
UASSER(foo == nullptr, "Should not nest");
VL_RESTORER(foo);
```

Reply:
Ok, but the macro should use UASSERT_OBJ.  Given that perhaps just put into the original code the UASSERT_OBJ to make sure there's no recursion and I'm satisfied.

Reply:
How about: If foo is subtype of AstNode (can figure it out via type_traits), use UASSERT_OBJ, otherwise use UASSERT?

Reply:
(otherwise need to pass an explicit obj to the macro)

Reply:
I think we need an explicit obj, as nearly always the first argument will be nullptr as it is what is being tracked after this call, and starts off null.

Reply:
Ok, I think I am making less and less sens actually, I will try to come up with something sensible.

Reply:
There is already a `UASSERT_OBJ(!m_senTreep, nodep, "Should not nest");` a the top of the function. I will leave this as is now and add the 3 operand VL_RESTORER at some later point to avoid bloating this patch.